src/plugins/responsive-preview/index.js

```diff
@@ -0,0 +1,190 @@
+							mediaIndices[ mediaIndex ] = {
+								mediaText: media.mediaText,
+								min,
+								max,
+								cssText,
+							}
+						} else {
+							// Store the cached value if nothing is modified.
+							mediaIndices[ mediaIndex ] = { ...cssObject[ index ][ mediaIndex ] }
+						}
+					} else {
+					// Store the new media query if the value is not found in cached media queries.
+					 mediaIndices[ mediaIndex ] = {
+							mediaText: media.mediaText,
+							min,
+							max,
+							cssText,
+					 }
+					}
+				}
+			} )
+		}
+
+		cssObject[ index ] = { ...mediaIndices }
+	} )
+}
+
+/**
+ * Updates the current media query based
+ * on Preview Mode.
+ *
+ * @param {string} previewMode the current Preview Mode of the editor.
+ * @param {number} width the editor's current width
+ */
+const updateMediaQuery = ( previewMode = 'Desktop', width = 0 ) => {
+	if ( previewMode === 'Tablet' || previewMode === 'Mobile' ) {
+		// If Preview is in Tablet or Mobile Mode, modify media queries for Tablet or Mobile.
+		keys( cssObject ).forEach( styleSheetIndex => {
+			keys( cssObject[ styleSheetIndex ] ).forEach( index => {
+				const {	min, max } = cssObject[ styleSheetIndex ][ index ]
+				if ( inRange( width, min, max ) ) {
+					if ( document.styleSheets[ styleSheetIndex ].cssRules[ index ].media.mediaText !== 'screen and (max-width: 5000px)' ) {
+						document.styleSheets[ styleSheetIndex ].cssRules[ index ].media.mediaText = 'screen and (max-width: 5000px)'
+					}
+				} else if ( document.styleSheets[ styleSheetIndex ].cssRules[ index ].media.mediaText !== 'screen and (min-width: 5000px)' ) {
+					document.styleSheets[ styleSheetIndex ].cssRules[ index ].media.mediaText = 'screen and (min-width: 5000px)'
+				}
+			} )
+		} )
+	} else {
+		// If Preview is in Desktop Mode, revert all media queries
+		keys( cssObject ).forEach( styleSheetIndex => {
+			keys( cssObject[ styleSheetIndex ] ).forEach( index => {
+				if ( document.styleSheets[ styleSheetIndex ].cssRules[ index ].media.mediaText !== cssObject[ styleSheetIndex ][ index ].mediaText ) {
+					document.styleSheets[ styleSheetIndex ].cssRules[ index ].media.mediaText = cssObject[ styleSheetIndex ][ index ].mediaText
+				}
+			} )
+		} )
+	}
+}
+
+const observerCallback = () => {
+	const {
+		__experimentalGetPreviewDeviceType: getPreviewDeviceType,
+	} = select( 'core/edit-post' )
+	const mode = getPreviewDeviceType()
+
+	// Only call when switching to desktop, or in responsive mode.
+	if ( mode === 'Desktop' && previousMode === 'Desktop' ) {
+		return
+	}
+	previousMode = mode
+
+	// Gets the current width of the visual editor
+	// Cache the width since it's expensive to perform getComputedStyle
+	if ( ! widthsDetected[ mode ] ) {
+		const visualEditorEl = document.querySelector( query )
+		widthsDetected[ mode ] = parseInt( window.getComputedStyle( visualEditorEl ).width, 10 )
+	}
+
+	// Populate the cssObject with media queries and cache values if necessary.
+	cacheCssObject()
```

Comment:
You may need to throttle this, since goes through a lot of styles for better performance.